### PR TITLE
Fix fmf_root when using dist-git

### PR DESCRIPTION
**Pull request by LecrisUT**
I don't think this is a suitable "fix", but it highlights the logic issue
- `path` is only used when `dist-git-extract` is specified
- `dist-git-extract` must be a directory in the current implementation. It does not behave as documented, i.e. to specify the sources to copy over
  One would expect in this case the root to be the same as if passed `dist-git-extract`, but it is actually flattened out
- `path`  should support glob pattern because the extracted source is generally `package_name-0.1.2`

TODO:
- [ ] Document the spec file change of `dist-git-extract` in help, stories and cli

### CODE REVIEW

File: tmt/steps/discover/fmf.py [hunk 1]
```diff
@@ -407,8 +407,12 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
                 else:
                     try:
-                        dist_git_extract = glob.glob(os.path.join(
-                            sourcedir, dist_git_extract.lstrip('/')))[0]
+                        # Glob everything
+                        glob_path = glob.glob(os.path.join(sourcedir, dist_git_extract.lstrip('/')))
+                        # Use only the directories
+                        glob_dirs = [d for d  in glob_path if Path(d).is_dir()]
+                        # Use only the first valid directory
+                        dist_git_extract = glob_dirs[0]
```
Comment (happz) [inline tmt/steps/discover/fmf.py:415]: `os.path.join` is a left-over from our recent transition to `pathlib.Path`, let's get rid of it :)

How about this?

```python
glob_dirs = [
    child
    for child in sourcedir.glob(dist_git_extract.lstrip('/'))
    # Use only the directories
    if child.is_dir()
]

# Use only the first valid directory
dist_git_extract = glob_dirs[0]
```

Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:415]: I've had a conference and a cold so I didn't work on this, but I think we should change the logic here:
- no `path` is passed use the search algorithm
- if `path` is passed it has to be relative to `BUILDROOT`
  - in this regard dist_git_extract should not take only the 1st glob_dirs, but copy all files and glob patterns
  - allow for globbing for the path, i.e. one should pass `scikit-build-core-*/path/to/fmf_root`

Comment (happz) [reply tmt/steps/discover/fmf.py:415]: TBH, the whole function is hard to navigate for me, therefore I can't comment too much on the change in its workflow. I was only after `os.path.join` in this case :) Let's ping @lukaszachy who `git blame`s for these lines.

**Review by lukaszachy — CHANGES_REQUESTED** [review]

**Comment by thrix** [comment]
@LecrisUT hi, any luck with this?

**Comment by LecrisUT** [comment]
Sorry I have been caught up a bit and didn't get around to work on this. I'll try to do it this weekend.
The main todo for this:
- [ ] Document the variables `dist-git-merge`, `dist-git-init`, `dist-git-extract`, `dist-git-remove-fmf-root` in stories
- [ ] Document the difference in the logic if `dist-git-extract` and/or `path` is specified or not in stories
       E.g. without `dist-git-extract` and `path`, it will default to use the first `fmf_root` it discovers, and use that as the root path moving on
- [ ] Change the `dist-git-extract` logic
       Currently when using a `dist-git-extract` it uses the extracted path as the new root, e.g. for calculating `path`. That is unintuitive and it complicates the patching.
       Instead, the whole paths defined in `dist-git-extract` should be copied as is specified, preserving the path structure as it would be inside the rpm spec file
- [x] When no `dist-git-extract` is specified (i.e. when using auto-detected fmf_root) completely ignore `path` variable
- [ ] If `dist-git-extract = /`, copy all extracted data and allow the usage of `path`
- [ ] Use `pathlib.Path` instead of `os.path.join`. I didn't get around to figuring out the context for that and where it needs to be changed.

If someone can help with the first 2, that would be great. I always have a hard time coming up with words. If you want to help with the other ones, let me know and go ahead and push to my branch. I can comment on later in the PR.

**Comment by LecrisUT** [comment]
Ok, I have completely reworked how `dist-git-extract`:
- `dist-git-extract` always copies all files specified in the glob pattern, preserving the folder structure.
- `dist-git-remove-fmf-root` is ambiguous where/how to remove the fmf_root. For now I have it removed
- `dist-git-init` now initializes at the root of the extracted path

This one now needs some reviewing that it is implemented correctly and there is no edge-case that should change the logic implemented. Also there needs more tests implemented.

File: tmt/steps/discover/fmf.py [hunk 2]
```diff
@@ -251,13 +244,15 @@ def go(self) -> None:
...
-        dist_git_remove_fmf_root = self.get('dist-git-remove-fmf-root', False)
 
         # Self checks
         if dist_git_source and not dist_git_merge and (ref or url):
             raise tmt.utils.DiscoverError(
                 "Cannot manipulate with dist-git without "
                 "the `--dist-git-merge` option.")
+        # normalize `dist_git_extract` to list of strings
+        if type(dist_git_extract) == str:
+            dist_git_extract = [dist_git_extract]
```
Comment (happz) [inline tmt/steps/discover/fmf.py:255]: I'd recommend using `tmt.utils.normalize_path_list()` (https://github.com/teemtee/tmt/blob/main/tmt/utils.py#L4383C5-L4383C24) right away - I forgot `discover` plugins do now use our `field()` implementation yet, therefore normalization is not yet "out of the box", but this function would be the best to take care of the input. It would give you a list of `Path` instances, no need for checking type of `dist_git_extract` before calling it.

File: tmt/steps/discover/fmf.py [hunk 3]
```diff
@@ -403,23 +398,40 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
 
             # Check what should be extracted from the sources
             if dist_git_extract:
-                if dist_git_extract == '/':
-                    dist_git_extract = sourcedir
+                if len(dist_git_extract) == 1 and dist_git_extract[0] == '/':
```
Comment (happz) [inline tmt/steps/discover/fmf.py:401]: After using `normalize_path_list()`, `dist_git_extract` would be a `List[Path]`. Comparing its element to string directly wouldn't work, `dist_git_extract[0].resolve() == '/'` should be the right way.

File: tmt/steps/discover/fmf.py [hunk 4]
```diff
@@ -403,23 +398,40 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
+                    # Special case to extract all sources
+                    dist_git_extract = [sourcedir]
                 else:
-                    try:
-                        dist_git_extract = glob.glob(os.path.join(
-                            sourcedir, dist_git_extract.lstrip('/')))[0]
-                    except IndexError:
+                    if any(p == '/' for p in dist_git_extract):
```
Comment (happz) [inline tmt/steps/discover/fmf.py:405]: Same as above, `p.resolve()` will be needed.

File: tmt/steps/discover/fmf.py [hunk 5]
```diff
@@ -403,23 +398,40 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
+                    glob_patterns = [str(sourcedir.joinpath(p.lstrip('/'))) for p in dist_git_extract if type(p) == str]
+                    # Remove all glob patterns and keep the dict objects
+                    dist_git_extract = [p for p in dist_git_extract if type(p) != str]
+                    for pattern in glob_patterns:
+                        # Add all resolved paths to the paths
+                        try:
+                            dist_git_extract.extend([{
+                                "src": Path(p),
+                                "dest": Path(p).relative_to(sourcedir),
+                            } for p in glob.glob(pattern)])
```
Comment (happz) [inline tmt/steps/discover/fmf.py:420]: This will be a problem: `dist_git_extract` is probably of incompatible type, with `normalize_path_list` it'd be `List[Path]`, therefore adding a dictionary to it would be disallowed. It looks like adding an accumulator variable for these objects would be better, plus also a dataclass to hold these `src`/`dst` pairs, or a simple tuple, `Tuple[Path, Path]` would be also fine. `actual_dist_git_extract: List[Tuple[Path, Path]]` would be a good fit - process `dist_git_extract`, create entries as needed and pushed them into the "normalized" accumulator.

File: tmt/steps/discover/fmf.py [hunk 6]
```diff
@@ -430,30 +442,30 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
-                if not dist_git_extract:
-                    dist_git_extract = sourcedir
-
-            # Now can safely default to top_fmf_root
+            # Set dist_git_extract if not set yet
             if not dist_git_extract:
-                dist_git_extract = top_fmf_root
+                dist_git_extract = [{
+                    "src": sourcedir,
+                    "dest": '.',
```
Comment (happz) [inline tmt/steps/discover/fmf.py:449]: Please, stick to `Path` as much as possible - `sourcedir` is probably already a `Path` instance, but instead of `'.'`, `Path.cwd()` would be much better. We try to force "path-like strings" to be represented as `Path` instances, for better control and linting.

File: tmt/steps/discover/fmf.py [hunk 7]
```diff
@@ -430,30 +442,30 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
 
             # Now copy dist_git_extract into tests
             if not self.opt('dry'):
-                tmt.utils.copytree(
-                    dist_git_extract,
-                    self.testdir,
-                    symlinks=True,
-                    dirs_exist_ok=True)
+                # Copy all requested paths
+                for path in dist_git_extract:
```
Comment (happz) [inline tmt/steps/discover/fmf.py:455]: But `path` is not actually a path, is it? It's a mapping, src/dst pair - but with my proposal of changing this to `Path, Path` tuples, we could do a nice `for src, dst in actual_dist_git_extract` expansion :)

File: tmt/steps/discover/fmf.py [hunk 8]
```diff
@@ -430,30 +442,30 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
...
+                    try:
+                        tmt.utils.copytree(
+                            path['src'],
+                            self.testdir.joinpath(path['dest']),
```
Comment (happz) [inline tmt/steps/discover/fmf.py:459]: `self.testdir / path['dest']` (or `dst` as suggested above).

File: tmt/steps/discover/fmf.py (hunk 5 above)
Comment (happz) [reply tmt/steps/discover/fmf.py:420]: Actually, a couple of more suggestions:

* `# Any string items are glob patterns, otherwise they are dicts with source and destination` - this does not seem to be correct, `dist_git_extract` seems to contain strings only, so far, the first dictionary is added to it in the `for pattern in glob_patterns` loop below. It's all strings now - or, better, it should be all `Path` instances after `normalize_path_list()` is applied at the beginning.
* `glob_patterns` is working with strings, but `Path` already offers `glob` support. How about something like this?

```python
for pattern in dist_git_extract:
    for path in sourcedir.glob(str(pattern)):
        actual_dist_git_extract.append((path, path.relative_to(sourcedir)))
```

This should consume patterns (`Path` instances), ask `sourcedir` to list matching paths, and add them all to the new accumulator as `Tuple[Path, Path]`.

File: tmt/steps/discover/fmf.py (hunk 2 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:255]: Woud it work with glob strings?

The other thing is that I have the accumulator and the glob list merged in the same input `dist-git-extract`. I feel it would be confusing for the user to have it separated as it might indicate exclusivity between the options, different default logics, etc. But in allowing for that, that function cannot be used right?

I think it's not a big deal for it to be actaul paths at that point because it will always go through glob pattern parsing.

File: tmt/steps/discover/fmf.py (hunk 5 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:420]: The key here is that the user can add dict items instead of just strings (see the spec file changes). A tuple would be incompatible with that spec file changes right?

File: tmt/steps/discover/fmf.py (hunk 6 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:449]: Oh indeed, I forgot `Path('.')`  does not expand itself. I need it in the form of `.` because I use it to `join_path` later on.

File: tmt/steps/discover/fmf.py (hunk 7 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:455]: Hmm indeed should use a different naming, but it should be consistent with the spec file format for simplicity

File: tmt/steps/discover/fmf.py (hunk 8 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:459]: Oh `/` is an alias of `joinpath`? Just making sure there is no weird things that can happen since the filereaders/loaders are overwiriding `joinpath` and `iterdir` when they are implementing custom readers like zip reader, lookaside, etc.

**Review by LecrisUT — COMMENTED** [review]
So one key point here is that the spec for `dist-git-extract` is changed to allow the user to manually overwrite the globbing logic and use `src-dest` dict items

File: tmt/steps/discover/fmf.py (hunk 8 above)
Comment (happz) [reply tmt/steps/discover/fmf.py:459]: Yep, `/` is an alias for `joinpath`, it works in the same way by the magic of dunder methods.

File: tmt/steps/discover/fmf.py (hunk 5 above)
Comment (happz) [reply tmt/steps/discover/fmf.py:420]: I see, then this needs more work, I'm afraid.

* specification now says:

  ```
  dist-git-extract
            directory (glob supported) to copy from extracted sources.
            Set to ``/`` to extract all sources"
   ```
   That does not mention any use of mappings - yes, *schema* has changed, but not the help text or documentation. It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.

* ok, let's say `dist-git-extract` can now accept also a mapping with `src` and `dst` keys - *from fmf*. How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.

Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.

Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice. I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:

```python
def normalize_dist_git_extract(
        key_address: str,
        value: Any,
        logger: tmt.log.Logger) -> List[Tuple[str, Optional[str]]]:
    if value is None:
        return []

    # Only a single string, must be a glob pattern
    if isinstance(value, str):
        return [(value, None)]

    if isinstance(value, (list, tuple)):
        normalized = []

        for item in value:
            if isinstance(item, dict):
                normalized.append((item['src'], item['dst']))

            elif isinstance(item, str):
                # TODO: how to deal with `--dist-git-extract ??`?
                # --dist-git-extract foo:bar => maybe if ':' in item: src, dst = item.split(':'); normalized.append((src, dst) ?

                normalized.append((item, None))

        return normalized

dist_git_extract = normalize_dist_git_extract(self.get('dist-git-extract', []))
```

2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:

```python
for src, dst in dist_git_extract:
    if dst is None:
        for path in sourcedir.glob(src):
            actual_dist_git_extract.append((path, path.relative_to(sourcedir)))
    else:
        actual_dist_git_extract.append((Path(src), Path(dst))
```

A mix of str and dicts turns into `List[Tuple[str, Optional[str]]]` when we first meet it, then it turns into a `List[Tuple[Path, Path]]` with patterns resolved, to comply with the "`Path` for path-like strings" rule. The transition is explicit, it would handle all kinds of input, and it's clear where a pattern or mapping turns into paths.

File: tmt/steps/discover/fmf.py (hunk 2 above)
Comment (happz) [reply tmt/steps/discover/fmf.py:255]: See https://github.com/teemtee/tmt/pull/2098/files#r1241278339, this will need more work. I missed the added `src/dst` support, normalization will need to be more complex.

File: tmt/steps/discover/fmf.py (hunk 5 above)
Comment (LecrisUT) [reply tmt/steps/discover/fmf.py:420]: Thanks for the tips. I've added a quick TODO on the top about the documentation issue. Feel free to update the top comment so that it's more relevant with the current status.

> I see, then this needs more work, I'm afraid.

No worries, it is to be expected with the spec changes

>  It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.

Yes, the latter is correct. I would prefer to allow glob-like path, but how can that be done for `dst`? Is there an equivalent to the regex group matcher that can be used to replace in the glob pattern?

>  How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.

Oh I did not consider the cli. Is there a reference for another dict-like object that is passible in the cli? I think the best is the latter implementation since it would not require any more post-processing? The former would parse it as tuple? The other nuance is how to change the cli to accept list parameters for this? I would have thought there is a mechanism to synchronize it with the provided json-spec.

> Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.

Understood. So we need to extract the normalizer to a separate function. Besides the `if string` normalization, is there anything else to be added? The `List[str|dict]` -> `List[dict]` can only happen in this file since it needs the extracted source. 

> Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice

You mean the `dict` -> `tuple`/`NamedTuple`? Sure I don't mind that. If the conversion to a NamedTuple can be done automatically on the importer side that would simplify things.

> I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:

I think this depends on if we can solve the globing support for `dst`, otherwise the logic of if it's `Path` or a tuple are rather different. I'll have to look at the section again and see if there is any potential confusion. E.g. what if the user only provided `src`, could it be misinterpreted and allowed through despite the ambiguity?

>     2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:

Good point I'll make a different variable

Comment (happz) [reply tmt/steps/discover/fmf.py:420]: > > It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.
> 
> Yes, the latter is correct. I would prefer to allow glob-like path, but how can that be done for `dst`? Is there an equivalent to the regex group matcher that can be used to replace in the glob pattern?

I suppose the glob pattern is fine, regex pattern would also work, but that can be resolved later.

> 
> > How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.
> 
> Oh I did not consider the cli. Is there a reference for another dict-like object that is passible in the cli? I think the best is the latter implementation since it would not require any more post-processing? The former would parse it as tuple? The other nuance is how to change the cli to accept list parameters for this? I would have thought there is a mechanism to synchronize it with the provided json-spec.

I'm afraid not, there is no out-of-the-box solution (yet). Options given by user would arrive as strings, or rather a list of strings, e.g. `--dist-git-extract foo:bar --dist-git-extract baz` would land in a normalize function as `['foo:bar', 'baz']`. Normalization must take care of all three types of inputs (simple string, dicts from fmf, and possibly src/dst pair in a single string from command line). My example above tries to address all three, although the CLI par form is left commented out as I was not sure how exactly would such an option look like. The important bit is, when normalization is over, we get a uniform list of items, representing the inputs.

> > Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.
> 
> Understood. So we need to extract the normalizer to a separate function. Besides the `if string` normalization, is there anything else to be added? The `List[str|dict]` -> `List[dict]` can only happen in this file since it needs the extracted source.

Handling of `--dist-git-extract src:dst` must happen here as well (or whatever the exact form would be, simply the CLI input must be parsed and converted here as well.)


> > Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice
> 
> You mean the `dict` -> `tuple`/`NamedTuple`? Sure I don't mind that. If the conversion to a NamedTuple can be done automatically on the importer side that would simplify things.

tmt has no out-of-the-box solution for this conversion, but it would happen in the normalization function, and that's the right place for it.

> > I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:
> 
> I think this depends on if we can solve the globing support for `dst`, otherwise the logic of if it's `Path` or a tuple are rather different. I'll have to look at the section again and see if there is any potential confusion. E.g. what if the user only provided `src`, could it be misinterpreted and allowed through despite the ambiguity?

I'd probably go with both or none, if there's just a single key, the normalization function performing the conversion from dicts into something more rigid, tuples or whatever, would raise a `tmt.utils.NormalizationError`, "hey, both keys are needed". But if there's clear semantics and use case for just `src`, it'd be fine too.

> > 2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:
> > ```
> 
> Good point I'll make a different variable

**Comment by psss** [comment]
Is this one still relevant? @lukaszachy, there were some recent changes in the dist-git implementation...

**Comment by therazix** [comment]
@lukaszachy Is this still relevant?

**Comment by LecrisUT** [comment]
The issue, yes, but the approach, no. To refresh on the original issue, you cannot specify where the tmt root is relative to the `dist-git`, leading to the need of a workaround like [^1]
```yaml
discover:
  how: fmf
  dist-git-source: true
  dist-git-extract: /
  path: spglib-2.6.0/test
```
Instead I would like it to be
```yaml
discover:
  how: fmf
  dist-git-source: true
  # OptionA
  path: test
  # OptionB
  path: spglib-*/test
```

The thing that gets in the way is if you don't specify `dist-git-extract`, it would try to go to the first fmf root it finds and copy the fmf tree. The `path` value at that point is meaningless, so `path` needs to be accounted beforehand
https://github.com/teemtee/tmt/blob/f3a0e53c633e727ded45cc7fbfd60f63b79c9487/tmt/steps/discover/fmf.py#L830-L840

But there are some other inconsistencies in the `post_dist_git` that randomly interfere.

[^1]: https://matrix.to/#/!GcyWdIZAIxEoRmGXjV:fedoraproject.org/$4NPsm9cPBPF4qgVn8EQmeEW991fgPSLtGn-uEwD2isA?via=fedoraproject.org&via=fedora.im&via=matrix.org

**Comment by lukaszachy** [comment]
Apologies for taking too much time and introducing the mess in the first place. At this time, I think the best would be to clarify expected behavior, as current implementation is buggy, as I tried too much to avoid yet another option and didn't realize there can be more TMT_SOURCE directories...

The main problem is the ambiguity in `--path` which can be used as workaround to try local changes as opposed to `--url`. 

Thus if the plan is set like:
```
discover:
  how: fmf
  dist-git-source: true
  url: https://src.fedoraproject.org/rpms/fmf
```

one could just swap `url:` with `path: /tmp/local/fmf` and try changes in the spec file (adding patches, etc.) without the need to push them to the remote. This is currenly broken and tmt requires setting the `dist-git-merge` option but the intention was like this.

Which brings the messy  `--dist-git-extract` which was supposed to be the 'path' inside the extracted rpm sources to copy in and use in the discovery. I'm not sure anymore why I thought flattenning was the good idea, to be really frank. Probably to skip copying makefile/binary sources. 

The `--dist-git-init` is a workaround to allow discovery if fmf root isn't part of the source tarball. The `--dist-git-remove-fmf-root` makes it possible to combine fmf trees. 

I'd like to keep way to try local changes, but it doesn't need to be via `path` option as help message is really about the root. A dedicated option (and PR) will be more appropriate.


### With all this background said, IMO use cases we need to support in tmt:

**UC1 - source tarball provides both tmt metadata and test sources**

Ideally `dist-git-source: true` is all one needs. tmt should be able to detect where the fmf root is (usually in ./tests, right) but `path` can be used for explicit location (glob supported). Only content of `path` (fmf root) is copied to the tests directory (respecting `prune` and `type: file` requires).

**UC1.1 - source tarball provides tmt metadata and test sources but is missing fmf root**

`dist-git-init` to create fmf root in directory specified by `path` option, otherwise same as UC1

**UC2 - source tarball provides only test sources, metadata provided by plan's repo**

`dist-git-source: true` and `dist-git-merge: true`.  
How to specify which files are to be copied to the tests directory? Everything unless `dist-git-extract` is set? 

**UC3 - source tarball provides both tmt metadata and test sources, user wants to use metadata from plan instead**

Partially solved by `dist-git-remove-fmf-root` but if there are multiple roots a better way is needed (#3155). Problem with specifying what to copy (possible job for dist-git-extract)

**Comment by LecrisUT** [comment]
> The main problem is the ambiguity in `--path` which can be used as workaround to try local changes as opposed to `--url`.

Oh, interesting point. Indeed that's an interesting workaround that would be useful for me also. Maybe we can minimize the impact here if we add only 1 variable `dist-git-extracted` and have that as the override for skipping all `dist-git` operations. Alternatively, wouldn't `tmt run discover` completely override the step so the special handling of `--path` is not really an issue?

The workflow I have in mind is that the user would do all the extraction, patching and everything they want in the `%prep`, including any `tmt init` and at that point, there is no need for the original `discover.how=fmf` step and it can be overwritten completely.

> ### With all this background said, IMO use cases we need to support in tmt:
> 
> **UC1 - source tarball provides both tmt metadata and test sources**
> 
> Ideally `dist-git-source: true` is all one needs. tmt should be able to detect where the fmf root is (usually in ./tests, right) but `path` can be used for explicit location (glob supported). Only content of `path` (fmf root) is copied to the tests directory (respecting `prune` and `type: file` requires).

Makes sense to me. Do we need it to be copied though? We already have the dist-git source available to all plan steps and further via `TMT_SOURCE_DIR` (although where this points is still ambiguous). Wouldn't it just work as if it was pointing to a different tmt tree?

My proposal on this is that `TMT_SOURCE_DIR` will contain everything as if `%prep` command was done and it would contain all the files under `BUILD/%{name}-%{version}-build` (i.e. you would usually have another `%{name}-%{version}` folder right under there). Then the `path` will be calculated as relative to `TMT_SOURCE_DIR` instead of relative to the current tree (there are actually other ambiguities of how `path` is calculated in other contexts that we should address/clarify at some point).

`path` may be written as a `glob` pattern, but it **must** expand to a single value
 
> **UC1.1 - source tarball provides tmt metadata and test sources but is missing fmf root**
> 
> `dist-git-init` to create fmf root in directory specified by `path` option, otherwise same as UC1

Could we also assume we do not have `.fmf` in dist-git in this situation? Just test sources.

> **UC2 - source tarball provides only test sources, metadata provided by plan's repo**
> 
> `dist-git-source: true` and `dist-git-merge: true`. How to specify which files are to be copied to the tests directory? Everything unless `dist-git-extract` is set?

In this case, I would expect files to be copied to the current tree. Here I would remove the `dist-git-extract` and replace it with `path` (aliasing for compatibility). `path` would have again the same semantics that it is relative to `TMT_SOURCE_DIR` and it accepts `glob` but must expand to a single value. The destination would *always* be the root of the current tree.

If the user needs more files, specific destination, or more complicated steps, they should use `prepare` + `TMT_SOURCE_DIR` instead, editing the `TMT_TREE`.

> **UC3 - source tarball provides both tmt metadata and test sources, user wants to use metadata from plan instead**
> 
> Partially solved by `dist-git-remove-fmf-root` but if there are multiple roots a better way is needed (#3155). Problem with specifying what to copy (possible job for dist-git-extract)

I think this should be covered by `plan.import` instead. It would be too confusing that a `discover` step influences a plan's metadata. My thought for this is that it would only be applied under the `$TMT_SOURCE_DIR/$path` directory and would just complement UC2, although it doesn't have much impact if you do not make the destination configurable. They could also achieve the same effect with manual `prepare` + `TMT_SOURCE_DIR` steps.

---

### Plan of action

Open a new clean PR and resolve the issue by:
1. Making sure `TMT_SOURCE_DIR` is available for all plan steps -> test steps
2. Decouple the `path` and make it relative to `TMT_SOURCE_DIR`
3. When `dist-git-merge` is set, use `path` as the "what to extract" again relative to `TMT_SOURCE_DIR`
4. Separate `dist-git-prepare` steps from the "re-run discover" step so that users have a hook to add steps in between these 2
5. Alias `dist-git-extract` to `path` and deprecate it
6. Deprecate `dist-git-remove-fmf-root` in favor of manual `prepare` steps

How does it sound?

**Comment by happz** [comment]
@lukaszachy please, check the https://github.com/teemtee/tmt/pull/2098#issuecomment-2883055478 comment, and share your thoughts on the proposed "plan of action".

**Comment by lukaszachy** [comment]
One big canned worm to decide first. Assuming `url` can be used to point to remote dist-git plan: Can we have plan using multiple dist-git configs?

Use case in mind: Integration of several components, among lines of

```
discover:
 - name: foo dependency
   how: fmf
   dist-git-source: true
   url: repo/foo.git
 - name: bar dependency
   how: fmf
   dist-git-source: true
   url: repo/bar.git
```

**Comment by lukaszachy** [comment]
> * Making sure `TMT_SOURCE_DIR` is available for all plan steps -> test steps

Ack, although it might be an array. Thankfully `$TMT_SOURCE_DIR` should keep working for the most common 'single item' use case.

> * Decouple the `path` and make it relative to `TMT_SOURCE_DIR`

Ack

 
> * When `dist-git-merge` is set, use `path` as the "what to extract" again relative to `TMT_SOURCE_DIR`

Ack
 
> * Separate `dist-git-prepare` steps from the "re-run discover" step so that users have a hook to add steps in between 
How will this be indicated on plan/CLI level? Seems very useful, not limited to dist-git. But it might be clashing with 'slicing' features, which happen in discover (right, @happz ?)

> * Alias `dist-git-extract` to `path` and deprecate it

Ack

> * Deprecate `dist-git-remove-fmf-root` in favor of manual `prepare` steps

Ack, it should be rare enough to let user do what they need, but ideally it will be just adding prepare.shell config to remove right files between 'rpmbuild -bp' and 're-run discover'.

**Comment by LecrisUT** [comment]
> One big canned worm to decide first. Assuming `url` can be used to point to remote dist-git plan: Can we have plan using multiple dist-git configs?

Sounds good, but out-of-scope rn. There needs to be more discussion, e.g. how do we deal with `TMT_SOURCE_DIR` in that case. For now we could enforce that there must be only one dist-git step until we gather some use-cases for this.

https://github.com/teemtee/tmt/issues/3802 is also relevant for how we would want to make the transition

**Comment by psss** [comment]
Summary from the hacking session discussion: Let's try to implement the new behaviour and allow users to enable it by opting in. The old implementation would stay with us for some time to allow fluent transition.

**Comment by therazix** [comment]
@LecrisUT Is this PR still relevant?
